### DISCUSSION TITLE
Systems that don't apparently boot with the latest ShredOS v2021.08.2_23_x86-64_0.34

### DISCUSSION THREAD

**Discussion by PartialVolume** [PartialVolume/shredos.x86_64#133] [Announcements]
I've started this discussion to show a good way of identifying the possible cause of ShredOS not booting on a specific computer system. This level of detail is really useful to me and helps to solve problems much quicker, without this detail i'm just guessing.

So, I just came across a laptop that would not boot the latest [v2021.08.2_23_x86-64_0.34](https://github.com/PartialVolume/shredos.x86_64/releases/tag/v2021.08.2_23_x86-64_0.34). It is a Packard Bell, Easynote TE69KB. It boots early versions of ShredOS that don't have the DRM & GPU drivers installed. All versions prior to this use a simple frame buffer with no graphics drivers.

Here's a screenshot of what I saw when it tries to boot ShredOS v2021.08.2_23_x86-64_0.34.
![boot screen-1](https://user-images.githubusercontent.com/22084881/226117050-091fd480-c705-4cef-b48f-311ca2ceb96c.jpg)
Now, it looks like ShredOS never fully booted, but in fact, ShredOS did boot you just can't see what happened when it transferred over from a frame buffer to the DRM graphics driver.

#### So how do we find out whether it really did boot and what might cause the display not to work?

Take the USB stick that you booted from and plug it into another system. Don't boot from it, all you want to do is edit a couple of files on it to enable ShredOS's build in telnet server.

1. Look for two grub.cfg files, you should edit both, adding telnetd=enable to the end of the linux boot command.

To enable the telnet server, edit /boot/grub/grub.cfg or/and /EFI/BOOT/grub.cfg and on the USB flash drive, add telnetd=enable to the kernel command line.

#### Example:
```
set default="0"
set timeout="0"

menuentry "shredos" {
	linux /boot/shredos console=tty3 loglevel=3 telnetd=enable
}
```

Umount the USB stick and return it to the non working system. ***Don't*** power cycle yet ..

2.  Grab a ethernet cable and connect your problem system into your local area network that should have DHCP enabled. If you don't know what DHCP is, it's most likely already enabled.

3. **Power cycle the system**. You will see the screen as per the picture above.

4. Now from another computer on the same network, run the following command to find the IP address of the non working system:

>nmap -p23 192.168.0/24 --open
Starting Nmap 7.80 ( https://nmap.org ) at 2023-03-18 15:22 GMT
Nmap scan report for **192.168.0.231**
Host is up (0.0035s latency).
.
PORT   STATE SERVICE
23/tcp open  telnet

Nmap has found a open telnet port on 192.168.0.231. So lets see if we can login:

```
telnet 192.168.0.231
Trying 192.168.0.231...
Connected to 192.168.0.231.
Escape character is '^]'.

shredos login: root
		{ no password }

sh-5.1# nwipe
```

We're in !. **So ShredOS is running.** 

5. The first port of call is the dmesg (diagnostic message buffer) this is the kernel's message buffer. There is most likely a clue in here as to what went wrong. Type `dmesg` and the log will be displayed. If you can't scroll back to the beginning then use `dmesg | more`. There is an advantage to just using dmesg without the more and that is the text is highlighted with different colors and the color that stands out is RED. So start by looking for any red text. If you want to save the dmesg text, type `dmesg > dmesg.txt` to write the dmesg output to the file dmesg.txt. You can then mount the USB flash drive and copy the dmesg.txt file to the USB flash drive with `cd /;mkdir shredos; mount /dev/sdx shredos; cp dmesg.txt /shredos/;umount shredos`. Replace /dev/sdx with the appropriate drive letter for the USB stick, /dev/sda or /dev/sdb etc. Use fdisk -l to list the drives if you're not sure.

6. So back to the Packard Bell. I followed the above procedure and this is what I found in the dmesg message buffer.

![Screenshot_20230318_162802](https://user-images.githubusercontent.com/22084881/226119266-0b6ec11c-dd6a-454c-8176-c2416051d67f.png)

Looks like the AMD  Kabini (Radeon HD8330) failed to start up properly due to some firmware issue.

7. I've now got a lead, so I can investigate what this issue is exactly. Could be the firmware is just missing from ShredOS and I need to install it.

>lspci
00:00.0 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 16h Processor Root Complex
**00:01.0 VGA compatible controller: Advanced Micro Devices, Inc. [AMD/ATI] Kabini [Radeon HD 8330]**
00:01.1 Audio device: Advanced Micro Devices, Inc. [AMD/ATI] Kabini HDMI/DP Audio
00:02.0 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 16h Processor Function 0
00:02.2 PCI bridge: Advanced Micro Devices, Inc. [AMD] Family 16h Processor Functions 5:1
00:02.3 PCI bridge: Advanced Micro Devices, Inc. [AMD] Family 16h Processor Functions 5:1
00:10.0 USB controller: Advanced Micro Devices, Inc. [AMD] FCH USB XHCI Controller (rev 01)
00:11.0 SATA controller: Advanced Micro Devices, Inc. [AMD] FCH SATA Controller [AHCI mode] (rev 40)
00:12.0 USB controller: Advanced Micro Devices, Inc. [AMD] FCH USB OHCI Controller (rev 39)
00:12.2 USB controller: Advanced Micro Devices, Inc. [AMD] FCH USB EHCI Controller (rev 39)
00:13.0 USB controller: Advanced Micro Devices, Inc. [AMD] FCH USB OHCI Controller (rev 39)
00:13.2 USB controller: Advanced Micro Devices, Inc. [AMD] FCH USB EHCI Controller (rev 39)
00:14.0 SMBus: Advanced Micro Devices, Inc. [AMD] FCH SMBus Controller (rev 3a)
00:14.2 Audio device: Advanced Micro Devices, Inc. [AMD] FCH Azalia Controller (rev 02)
00:14.3 ISA bridge: Advanced Micro Devices, Inc. [AMD] FCH LPC Bridge (rev 11)
00:18.0 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 16h Processor Function 0
00:18.1 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 16h Processor Function 1
00:18.2 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 16h Processor Function 2
00:18.3 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 16h Processor Function 3
00:18.4 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 16h Processor Function 4
00:18.5 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 16h Processor Function 5
01:00.0 Ethernet controller: Qualcomm Atheros QCA8171 Gigabit Ethernet (rev 13)
01:00.1 SD Host controller: Qualcomm Atheros Device 3010 (rev 13)
05:00.0 Network controller: Qualcomm Atheros QCA9565 / AR9565 Wireless Network Adapter (rev 01)

**Comment by PartialVolume** [comment]
A quick fix to the above problem of the Packard Bell laptop with AMD graphics not booting ShredOS.

I just added `nomodeset` to the linux kernel command line in both grub.cfg files and ShredOS then used the standard frame buffer as per previous versions of ShredOS and booted up without any issues.

**Comment by PartialVolume** [comment]
Quoting from the arch linux wiki regarding the use of nomodeset

>You may want to disable KMS for various reasons. To disable KMS, add nomodeset as a kernel parameter. See [Kernel parameters](https://wiki.archlinux.org/title/Kernel_parameters) for more info.
  Along with the nomodeset kernel parameter, for an Intel graphics card, you need to add i915.modeset=0, and for an Nvidia graphics card, you need to add nouveau.modeset=0. For Nvidia Optimus dual-graphics system, you need to add all the three kernel parameters (i.e. "nomodeset i915.modeset=0 nouveau.modeset=0").

**Comment by PartialVolume** [comment]
And the actual fix was the inclusion of the AMD firmware. I also added some extra video firmware which included in total

- AMD GPU video card firmware
- Intel i915 video card firmware
- AMD Radeon video card firmware
- Qualcomm Venus video codec accelerator
- Qualcomm Adreno GPU firmware 
- Samsung MFC video encoder/decoder v8 firmware

So I would imagine that fixes quite a few AMD based systems that would not boot using DRM.

![Added_AMD_firmware-small](https://user-images.githubusercontent.com/22084881/226147096-7f3128b7-4b4c-4ae2-a350-eefb14d7359c.jpg)

**Comment by PartialVolume** [comment]
Running dmesg and looking at the section that was previously failing, we can now see that the Kabini microcode now loads without any errors.

![Screenshot_20230319_003037](https://user-images.githubusercontent.com/22084881/226147354-210f1e57-9f7b-4a2e-a5cf-6abcad68a874.png)

**Comment by PartialVolume** [comment]
And now we are using DRM, the command `rtcwake -m mem -s 5` now works that allows the system to be put to sleep for 5 seconds then automatically rewake in order to unfreeze the drive so a ATA secure erase command can be issued to the drive, something that did not work when not using DRM or using the nomodeset setting.

So basically ready for nwipe's forthcoming ATA secure erase feature.

**Reply by Binary-Bear** [reply]
> So basically ready for nwipe's forthcoming ATA secure erase feature.

If there is time before the next release of nwipe would you please consider implementing some of the user requests in this [thread](https://github.com/PartialVolume/shredos.x86_64/discussions/74)?

Obviously I particularly like the features requested by Binary-Bear :)

This is also a [killer feature!](https://github.com/PartialVolume/shredos.x86_64/issues/69)

**Comment by PartialVolume** [comment]
Yes, i can add some of those features, HPA/DCO, ATA secure erase and PDF reports for each drive, but unlikely to have hot swap in a single instance of nwipe any time soon as due to the architecture of the code it would require a major rewrite.

However, I think you can probably do hot swap with this workaround, the key to this is that you must always run nwipe with a drive as an option. i.e nwipe /dev/sda, nwipe /dev/sdb. If you type nwipe on it's own while other nwipe's are running the nwipe you just tried to start will hang due to a 'feature' of a library we are using.

Here's eight nwipe's running at the same time, so you could wipe eight disks, as each nwipe completes and you exit it you can then swap the disc and start another wipe while all the others are still running.

tmux isn't currently in ShredOS but I could install it in the next version if anybody is interested in running multiple nwipe's at the same time ?

Eight nwipe's in tmux. 

![eight](https://user-images.githubusercontent.com/22084881/227073238-67f4008b-fe33-4bbe-b4ad-dca2b57aa7dd.png)

**Comment by Binary-Bear** [comment]
Thank you for adding the HPA/DCO, ATA secure erase and PDF reports for each drive.  If any of the other feature requests are simple for you to implement,or perhaps wouldn't take too much time, then including them is a great way to raise interest and re-energise a great project like ShredOS.

Again thank you for the suggestion of how to workaround a hot-swap.  It is a shame that the hot-swap ability would require a major rewrite as it would be a super feature.

I believe there will be many who would be interested in running multiple instances of nwipe at the same time.

Thank you PartialVolume for all the work you put into ShredOS.

**Reply by PartialVolume** [reply]
@Binary-Bear No problem, I enjoy the coding & producing code that is hopefully useful to others. 

Took me a little while to get tmux running in ShredOS today as tmux required utp-8 support which previously ShredOS did not have and it wasn't at first obvious how to add it. However UTF-8 is now supported.

For those bored moments when you start your disk wipes and realise it's going to take eleven hours to complete, here's a example of three nwipes running in the 2nd virtual terminal ( ALT F2 ) and something else running in the fourth pane to keep you occupied :-) This is running the latest development ShredOS (not uploaded until later today as a pre-release for testing).

ALT F2, loadkeys UK (or your country code, so the keyboard keys are correct, default is US), tmux, You need review tmux help but basically you need only the following commands CNTRL B ", CNTRL B %, CNTRL B {up/down/left/right} arrow keys. Those commands allow you create the 4 panes which hold the 4 terminals.

![test2](https://user-images.githubusercontent.com/22084881/227265034-d77b1103-d7cd-4f68-a72b-ac881095f1b1.png)

**Comment by Binary-Bear** [comment]
LOL that is great!

I should have mentioned in my previous requests that Space Invaders is far more important than hot-swap but you must have read my mind :)
